File: problems/famous_algorithms/topological_sort/js/topological_sort.js
export const topologicalSort = (() => {
  const State = {
    UNVISITED: 0,
    VISITING: 1,
    VISITED: 2,
  };

  const dfs = (job, graph, states, order) => {
    states.set(job, State.VISITING);

    const dependencies = graph.get(job) || [];
    for (const dep of dependencies) {
      if (states.get(dep) === State.VISITING) return false;
      else if (states.get(dep) === State.UNVISITED) {
        if (!dfs(dep, graph, states, order)) return false;
      }
    }

    states.set(job, State.VISITED);
    order.push(job);
    return true;
  };

  const initGraphStates = (graph, states, jobs) => {
    for (const job of jobs) {
      graph.set(job, []);
      states.set(job, State.UNVISITED);
    }
  };

  const buildGraph = (graph, deps) => {
    for (const [prereq, job] of deps) {
      if (!graph.has(prereq)) graph.set(prereq, []);
      graph.get(prereq).push(job);
    }
  };

  const topologicalSort = (jobs, deps) => {
    if (!Array.isArray(jobs) || !Array.isArray(deps)) {
      throw new Error("Invalid input: jobs and deps should be an array");
    }

    const graph = new Map();
    const states = new Map();
    const order = [];

    initGraphStates(graph, states, jobs);
    buildGraph(graph, deps);

    for (const job of jobs) {
      if (states.get(job) === State.UNVISITED) {
        if (!dfs(job, graph, states, order)) return []; // cycle detected
      }
    }
    return order.reverse();
  };

  return { topologicalSort };
})();

File: problems/famous_algorithms/topological_sort/cpp/topological_sort.cpp
#include "topological_sort.h"

bool dfs(int job, std::unordered_map<int, std::vector<int>> &graph,
         std::unordered_map<int, State> &states, std::stack<int> &order) {
  states[job] = VISITING;

  for (int dep : graph[job]) {
    if (states[dep] == VISITING)
      return false;
    else if (states[dep] == UNVISITED) {
      if (!dfs(dep, graph, states, order))
        return false;
    }
  }
  states[job] = VISITED;
  order.push(job);
  return true;
}

void initGraphStates(std::unordered_map<int, std::vector<int>> &graph,
                     std::unordered_map<int, State> &states,
                     const std::vector<int> &jobs) {
  for (int job : jobs) {
    graph[job] = std::vector<int>();
    states[job] = UNVISITED;
  }
}

void buildGraph(std::unordered_map<int, std::vector<int>> &graph,
                const std::vector<std::vector<int>> &deps) {

  for (const auto &dep : deps) {
    int prereq = dep[0];
    int job = dep[1];
    graph[prereq].push_back(job);
  }
}

std::vector<int> topologicalSort(std::vector<int> jobs,
                                 std::vector<std::vector<int>> deps) {
  std::unordered_map<int, std::vector<int>> graph;
  std::unordered_map<int, State> states;

  initGraphStates(graph, states, jobs);

  buildGraph(graph, deps);

  std::stack<int> order;
  for (int job : jobs) {
    if (states[job] == UNVISITED) {
      if (!dfs(job, graph, states, order))
        return {};
    }
  }
  std::vector<int> sortedJobs;
  while (!order.empty()) {
    sortedJobs.push_back(order.top());
    order.pop();
  }
  return sortedJobs;
}
